Dockerfile

```diff
@@ -1,24 +1,27 @@
-FROM node:10.15.3-alpine
-
-WORKDIR /usr/src/app
-RUN chown node:node /usr/src/app
-
+# Base stage installs production dependencies
+FROM node:10.15.3-alpine AS base
+ARG PORT=3000
+ENV PORT ${PORT}
 USER node
-
-# Install app dependencies
-# A wildcard is used to ensure both package.json AND package-lock.json are copied
-# where available (npm@5+)
+WORKDIR /home/node
 COPY --chown=node:node package*.json ./
+RUN npm ci
 
+# Development stage installs devDependencies, builds app from source and declares a file watcher as the default command
+FROM base AS development
+EXPOSE ${PORT} 9229 9230
 RUN npm install
-# If you are building your code for production
-# RUN npm ci --only=production
-
-# Bundle app source
-COPY --chown=node:node . .
+COPY --chown=node:node app/ /home/node/app/
 RUN npm run build
+CMD [ "npm", "run", "start:watch" ]
 
-ARG PORT=3000
-ENV PORT ${PORT}
-EXPOSE ${PORT} 9229 9230
-CMD [ "node", "index" ]
+# Test stage copied in Jest configuration and declares the test task as the default command
+FROM development AS test
+COPY --chown=node:node jest.config.js /home/node/jest.config.js
+CMD [ "npm", "run", "test" ]
+
+# Production stage exposes service port, copies in built app code and declares the Node app as the default command
+FROM base AS production
+EXPOSE ${PORT}
+COPY --from=builder /home/node/app/ /home/node/app/
+CMD [ "node", "app/index" ]
```

Comment:
As node is not designed to run as PID 1 this can cause issues in Docker containers when responding to interrupt signals. To avoid this this problem something like `tini` could be used as documented here https://github.com/krallin/tini#using-tini